sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessEvaluationContext.java

```diff
@@ -0,0 +1,429 @@
+    }
+  }
+
+  private void checkCallbacks(AppliedPTransform<?, ?, ?> producingTransform) {
+    TransformWatermarks watermarks = watermarkManager.getWatermarks(producingTransform);
+    ShouldFirePredicate shouldFire = new ShouldFirePredicate(watermarks.getOutputWatermark());
+    WatermarkCallback callback;
+    do {
+      callback = pendingCallbacks.pollIfSatisfies(producingTransform, shouldFire);
+      if (callback != null) {
+        callbackExecutor.execute(callback.getCallback());
+      }
+    } while (callback != null);
+  }
+
+  private static class ShouldFirePredicate
+      implements SerializableFunction<WatermarkCallback, Boolean> {
+    private final Instant currentInstant;
+
+    public ShouldFirePredicate(Instant currentInstant) {
+      this.currentInstant = currentInstant;
+    }
+
+    @Override
+    public Boolean apply(WatermarkCallback input) {
+      return input.shouldFire(currentInstant);
+    }
+  }
+
+  /**
+   * Create a {@link UncommittedBundle} for use by a source.
+   */
+  public <T> UncommittedBundle<T> createRootBundle(PCollection<T> output) {
+    return InProcessBundle.unkeyed(output);
+  }
+
+  /**
+   * Create a {@link UncommittedBundle} whose elements belong to the specified {@link
+   * PCollection}.
+   */
+  public <T> UncommittedBundle<T> createBundle(CommittedBundle<?> input, PCollection<T> output) {
+    return input.isKeyed()
+        ? InProcessBundle.keyed(output, input.getKey())
+        : InProcessBundle.unkeyed(output);
+  }
+
+  /**
+   * Create a {@link UncommittedBundle} with the specified keys at the specified step. For use by
+   * {@link InProcessGroupByKeyOnly} {@link PTransform PTransforms}.
+   */
+  public <T> UncommittedBundle<T> createKeyedBundle(
+      CommittedBundle<?> input, Object key, PCollection<T> output) {
+    return InProcessBundle.keyed(output, key);
+  }
+
+  /**
+   * Create a {@link PCollectionViewWriter}, whose elements will be used in the provided
+   * {@link PCollectionView}.
+   */
+  public <ElemT, ViewT> PCollectionViewWriter<ElemT, ViewT> createPCollectionViewWriter(
+      PCollection<Iterable<ElemT>> input, final PCollectionView<ViewT> output) {
+    return new PCollectionViewWriter<ElemT, ViewT>() {
+      @Override
+      public void add(Iterable<WindowedValue<ElemT>> values) {
+        sideInputContainer.write(output, values);
+      }
+    };
+  }
+
+  /**
+   * Execute the specified callback after the watermark for a {@link PValue} passes the point at
+   * which the trigger for the specified window (with the specified windowing strategy) must have
+   * fired from the perspective of that {@link PValue}, as specified by the value of
+   * {@link Trigger#getWatermarkThatGuaranteesFiring(BoundedWindow)} for the trigger of the
+   * {@link WindowingStrategy}. When the callback has fired, either values will have been produced
+   * for a key in that window, or the window is empty.
+   */
+  public void callAfterOutputMustHaveBeenProduced(
```

Comment:
Hmm...
callAfterWouldFire(...)?

Something about this method name and/or javadoc was confusing. Any way to make it clearer what this is doing?